### PR TITLE
fix: trying to bind to an outbound ip returns an empty list of port bindings

### PR DESCRIPTION
**Pull request by atzoum**
Keep calling docker inspect with a short delay until all port bindings are properly populated.

## Related Issue or Design Document

When trying to bind to an outbound interface, the first docker inspect call right after the container has been created doesn't return port bindings properly in the response. This causes an empty host/port to be calculated by the resource.

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md) and signed the CLA.
- [ ] I have referenced an issue containing the design document if my change introduces a new feature.
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security vulnerability. 
      If this pull request addresses a security vulnerability, 
      I confirm that I got approval (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added the necessary documentation within the code base (if appropriate).

### CODE REVIEW

File: dockertest.go [hunk 1]
```diff
@@ -491,6 +491,26 @@ func (d *Pool) RunWithOptions(opts *RunOptions, hcOpts ...func(*dc.HostConfig))
 	}, nil
 }
 
+// inspectContainerWithRetries will repeat the inspect call until the container has port bindings assigned.
+func (d *Pool) inspectContainerWithRetries(id string, retry int) (*dc.Container, error) {
+	c, err := d.Client.InspectContainer(id)
+	if err != nil {
+		return nil, err
+	}
+	const maxRetries = 10
```
Comment (atzoum) [inline dockertest.go:510]: **Note:** you may verify the existence of the bug by setting `maxRetries = 0`

**Comment by atzoum** [comment]
@aeneasr can you please review?

File: go.mod [hunk 2]
```diff
@@ -14,7 +14,7 @@ require (
 	github.com/lib/pq v1.10.9
 	github.com/moby/term v0.5.0
 	github.com/opencontainers/image-spec v1.1.0
-	github.com/opencontainers/runc v1.1.13
+	github.com/opencontainers/runc v1.1.14
```
Comment (alexandear) [inline go.mod]: I'm not sure we need upgrading runc version in this PR. It's the work for @dependabot (see #538)

File: dockertest.go [hunk 3]
```diff
@@ -491,6 +491,26 @@ func (d *Pool) RunWithOptions(opts *RunOptions, hcOpts ...func(*dc.HostConfig))
...
+	if retry > maxRetries {
+		return c, nil
+	}
+	// wait for port bindings to be assigned
+	for _, bindings := range c.NetworkSettings.Ports {
+		if len(bindings) == 0 {
+			time.Sleep(100 * time.Millisecond)
+			return d.inspectContainerWithRetries(id, retry+1)
```
Comment (alexandear) [inline dockertest.go]: Is it possible to avoid recursion?

File: go.mod (hunk 2 above)
Comment (atzoum) [reply go.mod]: iirc a check was failing, shall I revert?

Comment (atzoum) [reply go.mod]: reverted

File: dockertest.go (hunk 3 above)
Comment (atzoum) [reply dockertest.go]: done

File: go.mod (hunk 2 above)
Comment (atzoum) [reply go.mod]: @alexandear here is the failing check
https://github.com/ory/dockertest/actions/runs/11347631243/job/31559482556

Comment (alexandear) [reply go.mod]: The maintainer merges #538 and the fail in this branch will be gone.

**Review by aeneasr — APPROVED** [review]

File: dockertest.go [hunk 4]
```diff
@@ -505,6 +505,37 @@ func (d *Pool) RunWithOptions(opts *RunOptions, hcOpts ...func(*dc.HostConfig))
...
+			for _, bindings := range c.NetworkSettings.Ports {
+				if len(bindings) == 0 {
+					return true
+				}
+			}
+			return false
+		}(); !hasEmptyPortBindings {
+			return c, nil
+		}
+		retryNum++
```
Comment (alexandear) [inline dockertest.go:534]: Can be placed into the `for` [`PostStmt`](https://go.dev/ref/spec#For_clause).

Created the PR #544 to show this

**Pull request merged**